### PR TITLE
Adding coverage report for server-side tests using istanbul

### PR DESCRIPTION
**Pull request by lirantal**
Code coverage is now built-in to our project and grunt tasks!

meanjs-code-coverage-2

and further results:

meanjs-code-coverage-1

To run the code coverage simply execute:
```bash
grunt coverage
```

Addresses:
* https://github.com/meanjs/mean/pull/569
* https://github.com/meanjs/mean/pull/760

### CODE REVIEW

**Comment by lirantal** [comment]
@ilanbiala please go ahead and commit your coveralls to this PR

**Comment by mleanos** [comment]
LGTM. Really slick! I really like the initializing status :+1:

**Comment by lirantal** [comment]
lol cool

**Comment by ilanbiala** [comment]
@lirantal I won't be able to test anything until the weekend, so until then if you want to take a crack at it with Coveralls' guide, go for it.

**Comment by lirantal** [comment]
@ilanbiala so let's merge this PR and we can later create a new one to extend the local coverage we're running with an integration to coveralls.

**Comment by lirantal** [comment]
@codydaig @mleanos do you guys want to review this PR? I thought maybe you have some experience with code coverage tools

**Comment by lirantal** [comment]
@sielay @abhijt-zanak maybe you also have something to add?

**Comment by ilanbiala** [comment]
@lirantal I'd rather add it all at once. I don't think people are really gonna benefit from just having coverage in the terminal compared to having Coveralls running for them.

**Comment by lirantal** [comment]
@ilanbiala WDYT?
we must merge this first to `master` for coveralls.io to actually pick it up and show information about code coverage

File: README.md [hunk 1]
```diff
@@ -3,6 +3,7 @@
 [![Gitter](https://badges.gitter.im/Join Chat.svg)](https://gitter.im/meanjs/mean?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)
 [![Build Status](https://travis-ci.org/meanjs/mean.svg?branch=master)](https://travis-ci.org/meanjs/mean)
 [![Dependencies Status](https://david-dm.org/meanjs/mean.svg)](https://david-dm.org/meanjs/mean)
+[![Coverage Status](https://coveralls.io/repos/meanjs/mean/badge.svg?branch=master&service=github)](https://coveralls.io/github/meanjs/mean?branch=coveralls-code-coverage)
```
Comment (codydaig) [inline README.md]: @lirantal The .svg is displaying from the master branch but the link is going to the coveralls-code-coverage branch.

Comment (lirantal) [inline README.md]: thanks! that's probably a leftover from taking README update off of ilan's branch.

Comment (lirantal) [inline README.md]: there's also a missing `service=github` in the actual link?

Comment (codydaig) [inline README.md]: I'm not entirely sure about that one, I just noticed the odd ball branch.

Comment (lirantal) [inline README.md]: nm, looked at their website, it's not required

**Comment by mleanos** [comment]
@lirantal LGTM... tested and verified the coverage outputs the logs correctly. Everything looks good.

And *STILL* like the Initializing status :)

**Comment by lirantal** [comment]
lol @mleanos  :)
Awesome, thanks!

File: gruntfile.js [hunk 2]
```diff
@@ -176,6 +176,22 @@ module.exports = function (grunt) {
...
     },
+    mocha_istanbul: {
+      coverage: {
+        src: testAssets.tests.server,
+        options: {
+          print: 'detail',
+          coverage: true,
+          require: 'test.js',
+          coverageFolder: 'coverage',
+          reportFormats: ['cobertura','lcovonly'],
```
Comment (ilanbiala) [inline gruntfile.js]: What's cobertura?

File: gruntfile.js [hunk 3]
```diff
@@ -176,6 +176,22 @@ module.exports = function (grunt) {
...
+          check: {
+            lines: 40,
```
Comment (ilanbiala) [inline gruntfile.js]: Do we want these limits?

File: package.json [hunk 4]
```diff
@@ -30,6 +30,7 @@
     "connect-mongo": "~0.8.1",
     "consolidate": "~0.13.1",
     "cookie-parser": "^1.3.2",
+    "coveralls": "^2.11.4",
```
Comment (ilanbiala) [inline package.json]: should be a devDep no?

File: test.js [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+/**
+ * Module dependencies.
+ */
+var app, db, config;
+
+var path = require('path');
+var app = require(path.resolve('./config/lib/app'));
+
+app.init(function (app, db, config) {
+	console.log('Initialized test automation');
```
Comment (ilanbiala) [inline test.js]: indent

File: test.js [hunk 6]
```diff
@@ -0,0 +1,13 @@
+'use strict';
+
+/**
+ * Module dependencies.
+ */
+var app, db, config;
```
Comment (ilanbiala) [inline test.js]: Why are db and config set up here but they don't seem to be used anywhere.

**Comment by lirantal** [comment]
@ilanbiala per your questions:
1. Cobertura is a code coverage tool which is commonly used in Java project. the grunt-mocha-istanbul plugin supports exports to the cobertura reporter format to integrate with hudson/jenkins. We use in our enterprise both Coberutra and Jenkins so I figure others will find this useful too without requiring extra tweaks.
2. About the check limits - for now we should keep that threshold because we don't have much more than that covered, so until we have more code coverage let's keep it on a 'sane' threshold. (I don't see a reason to keep a 90 threshold to show the project as not reaching it's goal and keeping that for who knows how long...
3. Moved coveralls to devDep
4. Corrected indentations
5. Fixed

**Comment by codydaig** [comment]
LGTM

**Comment by ilanbiala** [comment]
LGTM, @lirantal do we already see some coverage reports for this PR or not until we merge?

**Comment by lirantal** [comment]
After we'll merge (I'll do it later), and then we can also tweak and take care of areas for more test as we wish.

**Comment by ilanbiala** [comment]
Sounds good.

**Comment by ilanbiala** [comment]
@lirantal are we merging this or is there something left to add?

**Comment by lirantal** [comment]
Merging! :)

**Comment by ilanbiala** [comment]
@lirantal Coveralls still doesn't show anything https://coveralls.io/github/meanjs/mean when I log in and see the repo status. Is there something else we have to add?

**Comment by lirantal** [comment]
I'll check this

**Comment by lirantal** [comment]
@ilanbiala ok so I put a specific grunt task called coverage and we aren't calling it as part of the travis ci tests today, and I don't think we should. Instead, we should add another task to the Travis CI to run the coverage. I'll work it out and let you know.

**Comment by ilanbiala** [comment]
@lirantal :+1:

**Pull request merged**